VideoLocker/src/main/java/org/edx/mobile/http/HttpRequestDelegate.java

```diff
@@ -48,19 +50,24 @@ public T fromJson(String json) throws Exception{
 
     public T fetchData(boolean preferCache) throws Exception{
         String json = null;
-        String cacheKey = getCacheKey();
+        String cacheKey = endPoint.getCacheKey();
         if ( preferCache  || !NetworkUtil.isConnected(MainApplication.instance()) ){
             try {
                 json = cacheManager.get(cacheKey);
             } catch (Exception e) {
                 logger.error(e);
             }
-            if ( json != null )
+            if ( json != null ) {
                 try {
                     return fromJson(json);
                 } catch (Exception e) {
                     logger.error(e);
                 }
+            }
+        }
+        //if it is on the UI thread, we just can not make http call
+        if (Looper.getMainLooper().getThread() == Thread.currentThread()){
```

Comment:
@aleffert  yeah, agree, we should through exception, as it is not supposed to be called in the UI thread. 
the problem is with the overloaded roles for all those methods.  1) it can return the data from cache without service call, or 2) it makes the server side call..   for the case 1, if we provide async api interface, the coding will be complicated. 

let me decompose the method into two.